### ISSUE TITLE
With Air51.1.1.3 I can no longer open CustomSQLConnection databases

### ISSUE THREAD

**Issue by Devin-Licastro** [airsdk/Adobe-Runtime-Support#3438]
My app stores important information in a SQL database (flash.data.SQLConnection) located in in applicationStorageDirectory.

This previously worked fine with Air 50.2.4.3 however after updating to Air 51.1.1.3 the databases can no longer be opened, and show message: Error #3138: File opened is not a database file.

Remaking the database works and I can use it for that session, but on the next load I can not reopen them and the error appears again. 

After some testing I found the databases can still be opened repeatedly if they arent encrypted:
dbConnection = new SQLConnection();
dbConnection.openAsync(dbFile);

However, with encryption they can not be reopened a 2nd time:
dbConnection = new SQLConnection();
dbConnection.openAsync(dbFile, SQLMode.CREATE, null, false, 1024, encryptionKey);
-> Error #3138: File opened is not a database file.

Likely related, it seems that also the EncryptedLocalStore is no longer functioning in a similar way? Values can only be stored for the current session, and the next time the app opens they are all reset. 

Has something changes here? 
So far I have just tested this on Windows. 

I need to update to 51.1.1.3 soon due to the iOS Chinese language issue so this is making me a bit nervous.

**Comment by ajwfrost** [comment]
Hi

We will check this one -> the SQL database functionality should be completely independent of the EncryptedLocalStore functionality, but this does seem to be suspicious!

thanks

**Comment by itlancer** [comment]
@ajwfrost
May be it related to https://github.com/airsdk/Adobe-Runtime-Support/issues/3394

**Comment by ajwfrost** [comment]
@itlancer I'd forgotten to update that thread - will do that now - but that issue was iOS-specific and this seems to be an issue on Windows (too?)

**Comment by ajwfrost** [comment]
Yes I almost got caught by the same :-) 

@Devin-Licastro we're just trying a basic test case here but it's still working with 51.1.1.3... are you able to create a simple/standalone test case that we can run to reproduce this?

thanks

**Comment by Devin-Licastro** [comment]
Ah sorry for the confusion with the iOS comment guys.
 
So it turns out the issue appears to be that something behind the scenes has changed how com.adobe.air.crypto.EncryptionKeyGenerator works.

**This works on air 50, but not air 51:**

```
var encryptionSalt:String = "SomeSecureEncryptionValueForTheUser";
var keyGenerator:EncryptionKeyGenerator = new EncryptionKeyGenerator(); 
encryptionKey = keyGenerator.getEncryptionKey(encryptionSalt);
dbConnection.openAsync(dbFile, SQLMode.CREATE, null, false, 1024, encryptionKey);
```


**But this works on air 51:**

```
var byteArray:ByteArray = new ByteArray();
byteArray.writeUTFBytes("1234567890123456"); 
dbConnection.openAsync(dbFile, SQLMode.CREATE, null, false, 1024, byteArray);
```

So I guess as a workaround I can just use a different function to turn the user's database key into a 16 char encryption key before opening the database... maybe like a MD5.hash of the value and then just set the length to 16 or something.

For the EncryptedLocalStore issue, it just seems like no values are being saved or reloaded for the 2nd session, I saw there were some changes to ELS in the release notes. Is it a known issue that it isn't working the same on Windows or do I need to make any updates here?

**Comment by ajwfrost** [comment]
Ah - so yes, that EncryptionKeyGenerator stores the keys in ELS...
https://github.com/mikechambers/as3corelib/blob/master/src/com/adobe/air/crypto/EncryptionKeyGenerator.as
```
		 * @param overrideSaltELSKey	The EncryptionKeyGenerator creates and stores a random value 
		 * 								(known as a <i>salt</i>) as part of the process of 
		 * 								generating the encryption key. The first time an application 
		 * 								calls the <code>getEncryptionKey()</code> method, the salt 
		 * 								value is created and stored in the AIR application's encrypted 
		 * 								local store (ELS). From then on, the salt value is loaded from the 
		 * 								ELS.
		 * 								<p>If you wish to provide a custom String ELS key for storing 
		 * 								the salt value, specify a value for the <code>overrideSaltELSKey</code>
		 * 								parameter. If the parameter is <code>null</code> (the default) 
		 * 								a default key name is used.</p>
```

We have had issues with ELS not remembering data - will just double-check that on Windows too... so I expect this is the culprit, will double-check this..

**Comment by ajwfrost** [comment]
Yes this had an underlying problem with ELS - and doubly so on Windows, because of a change we'd made in the file locations which then got blocked by internal AIR security... -> will be a fix coming out soon in 51.1.1.4 (next week - although currently there are some build server issues here..)
